app/controllers/coronavirus_form/address_lookup_controller.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class CoronavirusForm::AddressLookupController < ApplicationController
+  def submit
+    redirect_to check_your_answers_url if session[:check_answers_seen]
+
+    if params[:postcode].blank?
+      flash.now[:validation] = [{ text: I18n.t("coronavirus_form.questions.address_lookup.hint"), field: "postcode" }]
+      render controller_path, status: :unprocessable_entity
+    elsif params[:address_string].present?
+      update_params
+      redirect_to support_address_path
+    else
+      begin
+        results = helpers.postcode_lookup(params[:postcode])
+        @addresses = generate_address_strings(results)
+        @postcode = params[:postcode]
+        render controller_path, status: :ok
+      rescue AddressLookupError => e
+        flash.now[:validation] = [{ text: e.message, field: "postcode" }]
+        render controller_path, status: :unprocessable_entity
+      end
+    end
+  end
+
+private
+
+  def update_params
+    uprn_and_address = params[:address_string].split("|")
+    address = uprn_and_address.last.split(",")
+
+    session[:support_address] = {
+      uprn: strip_tags(uprn_and_address.first&.strip).presence,
+      building_and_street_line_1: strip_tags(address[0]&.strip).presence,
+      building_and_street_line_2: strip_tags(address[1]&.strip).presence,
+      town_city: strip_tags(address[2]&.strip).presence,
+      county: strip_tags(address[3]&.strip).presence,
+      postcode: strip_tags(address[4]&.gsub(/[[:space:]]+/, "")).presence,
+    }
+  end
+
+  def generate_address_strings(results)
+    results.map do |address|
+      {
+        text: address.first.last,
+        value: "#{address.first.first} | #{address.first.last}",
+      }
+    end
+  end
```

Comment:
I think it'd be better if the API / helper class you have could provide the data in a format the controller requires, rather than putting extra logic into the controller. Ideally controllers are dumb and don't do anything other than call other classes

Reply:
Done.  We know use the fields from the API response directly.